examples/src/main/java/kafka/examples/ExactlyOnceMessageProcessor.java

```diff
@@ -215,6 +225,33 @@ private long getRemainingRecords(KafkaConsumer<Integer, String> consumer) {
         }).sum();
     }
 
+    private int retry(int retries, KafkaConsumer<Integer, String> consumer, ConsumerRecords<Integer, String> records) {
+        retries++;
+        if (retries > 0 && retries <= MAX_RETRIES) {
+            // retry: reset fetch offset
+            // the consumer fetch position needs to be restored to the committed offset before the transaction started
+            Map<TopicPartition, OffsetAndMetadata> committed = consumer.committed(consumer.assignment());
+            consumer.assignment().forEach(tp -> {
+                OffsetAndMetadata offsetAndMetadata = committed.get(tp);
+                if (offsetAndMetadata != null) {
+                    consumer.seek(tp, offsetAndMetadata.offset());
+                } else {
+                    consumer.seekToBeginning(Collections.singleton(tp));
+                }
+            });
+        } else if (retries > MAX_RETRIES) {
+            // continue: skip bad records
+            // in addition to logging, you may want to send these records to a DLQ for further processing
+            records.forEach(record -> {
+                Utils.printErr("Skipping record after %d retries: %s", MAX_RETRIES, record.value());
+                consumer.seek(new TopicPartition(record.topic(), record.partition()), record.offset() + 1);
+                consumer.commitSync();
+            });
+            retries = 0;
+        }
```

Comment:
I think we should also have a `else` case to print some error messages for retries < 0's case.

Reply:
I don't see how retries can be less than zero.

Reply:
Yes, I know, so if that happened, it should be an fatal error, right? Otherwise, we don't need to check `retries > 0` at the first `if` condition. WDYT?

Reply:
ok, let's add this check